File: Courses/PHP Web Basics - май 2019/18. Exercise Polymorphism, Interfaces and Abstraction/01.04.TelephoneFactory/Smartphone.php
<?php

include_once "BrowsInterface.php";
include_once "CallInterface.php";

class Smartphone implements CallInterface, BrowsInterface
{

    /**
     * @param string $url
     * @return string
     * @throws \Exception
     */
    public function browse(string $url): string
    {
        if (preg_match("/\\d/", $url)) {
            throw new \Exception("Invalid URL!");
        }
        return "Browsing: {$url}!";
    }
    /**
     * @param string $number
     * @return string
     * @throws \Exception
     */
    public function call(string $number): string
    {
        if (!preg_match("/^[\\d]+$/", $number)) {
            throw new \Exception("Invalid number!");
        }
        return "Calling... {$number}";
    }
}

File: Courses/C# OOP/12. Interfaces and Abstraction Exercise/Interfaces and Abstraction - Exercise/04.Telephony/Smartphone.cs
using _04.Telephony.Interfaces;
using System;
using System.Collections.Generic;
using System.Text;

namespace _04.Telephony
{
    class Smartphone : ICallable, IBrowsable
    {
        public string Browse(string url)
        {
            ValidateURL(url);
            return $"Browsing: {url}!";
        }

        public string Call(string phoneNumber)
        {
            ValidatePhoneNumber(phoneNumber);
            return $"Calling... {phoneNumber}";
        }

        private void ValidatePhoneNumber(string phoneNumber)
        {
            foreach (var c in phoneNumber)
            {
                if (!char.IsDigit(c))
                {
                    throw new ArgumentException("Invalid number!");
                }
            }
        }

        private void ValidateURL(string url)
        {
            foreach (var c in url)
            {
                if (char.IsDigit(c))
                {
                    throw new ArgumentException("Invalid URL!");
                }
            }
        }
    }
}
